### DISCUSSION TITLE
Bind group  allocation and garbage collection

### DISCUSSION THREAD

**Discussion by liujing7256** [gpuweb/gpuweb#4090] [Q&A]
Guys, I need some help!
 
I build a web render engine with webgpu as the backend ,  I found some issues ,  It will create new texture as attachement  every frame , the bind group memory is growing larger, it's not going to decrease by javascript garbage collection. Is there a way to explicitly manage bind group allocation ? like driver , texture , buffer , there is a destroy interface for user control ?

Thanks a lot ！

**Comment by greggman** [comment]
You should not need to create a texture every frame or a bind group every frame. The only things that need to be created every frame are command buffers and encoders (and maybe texture views for the canvas). 

I'm pretty sure if you check any of [these examples](https://webgpufundamentals.org/webgpu/lessons/webgpu-resources.html#examples), none of them are creating textures or bind groups every frame. I'd guess the same is true of [three.js](https://threejs.org/examples/?q=webgpu), [babylon.js](https://doc.babylonjs.com/setup/support/webGPU), [play canvas](https://github.com/playcanvas/engine/issues/3536), etc...

**Reply by liujing7256** [reply]
Thanks for your feedback!

I use framegraph to implement some post-processing effects,  so create new textures every frame , I will try to write some caching policies to reduce the creation of bindgroups.

Maybe bindgroup alloction and garbage collection need to improve,  allocate a lot of descriptors at once and then delete them, the memory footprint is also large, and it doesn't seem to be free later.

**Reply by greggman** [reply]
>  use framegraph to implement some post-processing effects, so create new textures every frame

Maybe you can cache those too? I'm not sure why post-processing would require new textures every frame.

you might find [this library](https://github.com/greggman/webgpu-memory) useful.

Checking [this example](https://threejs.org/examples/webgpu_backdrop.html) using that library [shows no allocations after it starts](https://jsgist.org/?src=e72cbce480272ae107041ff0da98ae4c). If you resize the window then new bind groups are created (probably for the post processing) but if you keep resizing you'll eventually see them get GCed , or sometimes if you wait a few seconds you'll see the bind groups number drop.

**Reply by liujing7256** [reply]
Thank you very much. I will study it more carefully. Maybe there is something wrong with the way I use it.
